intermediates-tutorial-update (#565)
smartcontractkit/documentation: docs/Introduction/tutorials/intermediates-tutorial.md

uint256 private s_fee;
```

 

> For the contract to keep track of addresses that roll the dice, the contract will need to use mappings. 
> [Mappings](https://medium.com/upstate-interactive/mappings-in-solidity-explained-in-under-two-minutes-ecba88aff96e) are unique `key => value` pair data structures similar to hash tables in Java.

Reviewer: ```suggestion
To keep track of addresses that roll the dice, the contract uses mappings. [Mappings](https://medium.com/upstate-interactive/mappings-in-solidity-explained-in-under-two-minutes-ecba88aff96e) are unique key-value pair data structures similar to hash tables in Java.
```

uint256 private s_fee;
```

> Below are the values for Kovan testnet network. They will be updated in the `constructor`.

Reviewer 2: ```suggestion
```

> Bellow are the values for Kovan testnet network. They will be updated in the next step, the `constructor`.

Reviewer 2: ```suggestion
Below are the values for Kovan testnet network. They will be updated in the `constructor`.
```

Reviewer 2: Also, would it make sense to move information about values for Kovan testnet network to the next section? It may meld better since we introduce the constructor in the next section as well

> - Key Hash:   0x6c3699283bda56ad74f6b855546325b68d482e983852a7a82979cc4807b641f4

Reviewer 2: ```suggestion
```

> - Fee:        0.1 LINK (100000000000000000)

Reviewer 2: ```suggestion
```

## Initializing the contract

The fee and the key hash must be initialized in the constructor of the contract. To use `VRFConsumerBase` properly, you must also pass certain values into its constructor. Below are the values for Kovan testnet network. They will be updated in the `constructor`.

- Key Hash:   0x6c3699283bda56ad74f6b855546325b68d482e983852a7a82979cc4807b641f4
- Fee:        0.1 LINK (100000000000000000)

> Only the owner (the address which create the smart contract) will be able to do some tasks. In order to do this, you will import a contract named [`ConfirmedOwner`](https://github.com/smartcontractkit/chainlink/blob/master/contracts/src/v0.8/ConfirmedOwner.sol) and use it to extend the contract that you create.

Reviewer: ```suggestion
The address that creates the smart contract is the owner of the contract. Only the owner is allowed to do some tasks. Import a contract named [`ConfirmedOwner`](https://github.com/smartcontractkit/chainlink/blob/master/contracts/src/v0.8/ConfirmedOwner.sol) and use it to extend the contract that you create.
```

> The fee and the key hash must be initialized in the constructor of the contract. To use `VRFConsumerBase` properly, you must also pass certain values into its constructor.

Reviewer 2: ```suggestion
The fee and the key hash must be initialized in the constructor of the contract. To use `VRFConsumerBase` properly, you must also pass certain values into its constructor. Below are the values for Kovan testnet network. They will be updated in the `constructor`.

- Key Hash:   0x6c3699283bda56ad74f6b855546325b68d482e983852a7a82979cc4807b641f4
- Fee:        0.1 LINK (100000000000000000)
```

4. Store the `requestId` and roller address.
5. Emit an event to signal that the die is rolling.

You must add a `ROLL_IN_PROGRESS` constant to signify that the die has been rolled but the result is not yet returned. Also add a `DiceRolled` event to the contract.

> Only owner can execute the `rollDice` function.

Reviewer 2: ```suggestion
Only the owner of the contract can execute the `rollDice` function.
```

You will now deploy your completed contract. This deployment is slightly different than the example in the [Deploy Your First Contract](/docs/deploy-your-first-contract/) guide. In our case, you will have to pass in parameters to the constructor upon deployment.

> Once compiled, you'll see a dropdown list that looks like this in the deploy pane:

Reviewer 2: ```suggestion
Once compiled, you'll see a dropdown menu that looks like this in the deploy pane:
```

![Remix contract selected](/files/intermediates-tutorial-01.png)

> Select the contract `VRFD20` (or the name you given to yours).

Reviewer 2: ```suggestion
Select the contract `VRFD20` (or the name you've given your contract).
```

> You will deploy it on Kovan testnet network.

Reviewer 2: ```suggestion
You will deploy this contract on the Kovan testnet network.
```

**Note**: You should add the corresponding LINK token to your MetaMask account first:
![Metamask Add Tokens Screens](/images/contract-devs/metamask-1.png)

If you encounter any issues, make sure to check you copied the address of the correct network:
![Metamask Verify Contracts Screen](/images/contract-devs/metamask-2.png)

Use your Metamask address on the Kovan network to request LINK and send 1 LINK to the contract address. This address can be found in Remix under **Deployed Contracts** on the bottom left.

![Remix contract address](/files/intermediates-tutorial-04.png)

# 7. How do I test `rollDice`?

After you open the deployed contract tab in the bottom left, the function buttons are available. Find `rollDice` and click the caret to expand the parameter fields. Enter your Metamask address, and click 'roll'.

You will have to wait a few minutes for your transaction to confirm and the response to be sent back. You can get your house by clicking the `house` function button with your address. Once the response has been sent back, you'll be assigned a *Game of Thrones* house!

> Maybe you notice that there are more buttons (functions) than you built in the smart contract. These were inherited from the smart contracts `VRFConsumerBase` and `ConfirmedOwner`, which were used in the contract definition.

Reviewer 2: ```suggestion
You may notice there are more buttons (functions) than you originally built in the smart contract. These were inherited from the smart contracts `VRFConsumerBase` and `ConfirmedOwner` which are used in the contract definition.
```
